File: src/kagan/core/_event_rendering.py
"""Semantic event rendering — shared protocol for all Kagan clients.

Transforms raw ``(event_type, payload)`` pairs into a client-agnostic
``RenderableEvent`` that carries *kind*, *title*, *body*, and *severity*.
Clients (TUI, web, VS Code, CLI chat) switch on ``RenderableEvent.kind``
instead of repeating event-type mapping logic.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any

# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------


class RenderableKind(StrEnum):
    TEXT = "text"
    THOUGHT = "thought"
    TOOL_START = "tool_start"
    TOOL_UPDATE = "tool_update"
    STATUS_CHANGE = "status_change"
    VERDICT = "verdict"
    NOTE = "note"
    ERROR = "error"
    MERGE = "merge"
    PLAN = "plan"


class Severity(StrEnum):
    INFO = "info"
    SUCCESS = "success"
    WARNING = "warning"
    ERROR = "error"


# ---------------------------------------------------------------------------
# Renderable event
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class RenderableEvent:
    kind: RenderableKind
    title: str
    body: str = ""
    severity: Severity = Severity.INFO
    metadata: dict[str, Any] = field(default_factory=dict)
    event_id: str = ""
    session_id: str = ""


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _acp_payload(payload: dict[str, Any]) -> dict[str, Any]:
    """Extract the nested ``payload.acp`` dict if present."""
    nested = payload.get("acp")
    if isinstance(nested, dict):
        return nested
    return {}


def _format_tool_name(raw: str) -> str:
    """Human-readable tool name.

    - ``mcp__kagan__task_get`` -> ``kagan / task_get``
    - ``toolu_abc`` / ``call_abc`` -> ``tool call``
    - ``functions__name`` -> ``functions / name``
    - ``snake_case`` -> ``snake case``
    """
    if raw.startswith("toolu_") or raw.startswith("call_"):
        return "tool call"
    if "__" in raw:
        parts = raw.split("__")
        if parts[0] in ("mcp", "functions") and len(parts) >= 3:
            return " / ".join(parts[1:])
        return " / ".join(parts)
    return raw.replace("_", " ")


def _extract_tool_title(payload: dict[str, Any]) -> str:
    """Extract and format the human-readable tool title from a payload."""
    acp = _acp_payload(payload)
    raw = str(
        acp.get("toolName")
        or acp.get("name")
        or acp.get("title")
        or payload.get("tool_name")
        or payload.get("toolName")
        or payload.get("name")
        or payload.get("tool_call_id")
        or payload.get("toolCallId")
        or payload.get("id")
        or "tool call"
    )
    return _format_tool_name(raw)


def _extract_tool_status(payload: dict[str, Any], fallback: str = "done") -> str:
    """Extract tool execution status from a payload."""
    acp = _acp_payload(payload)
    return str(acp.get("status") or payload.get("status") or fallback)


# ---------------------------------------------------------------------------
# Main render function
# ---------------------------------------------------------------------------


def render_event(
    event_type: str,
    payload: dict[str, Any],
    event_id: str = "",
    session_id: str = "",
) -> RenderableEvent | None:
    """Map a raw event into a :class:`RenderableEvent`.

    Returns ``None`` for events that should be silently skipped (e.g. a
    ``TOOL_CALL_UPDATE`` whose status is ``"completed"`` or ``"done"``).
    """

    if event_type == "OUTPUT_CHUNK":
        text = str(payload.get("text") or "")
        if not text:
            return None
        thought = bool(payload.get("thought"))
        kind = RenderableKind.THOUGHT if thought else RenderableKind.TEXT
        return RenderableEvent(
            kind=kind,
            title="Thinking" if thought else "Output",
            body=text,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "TOOL_CALL_START":
        title = _extract_tool_title(payload)
        return RenderableEvent(
            kind=RenderableKind.TOOL_START,
            title=title,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "TOOL_CALL_UPDATE":
        status = _extract_tool_status(payload, "done")
        if status in ("completed", "done"):
            return None
        title = _extract_tool_title(payload)
        return RenderableEvent(
            kind=RenderableKind.TOOL_UPDATE,
            title=title,
            body=status,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "AGENT_STATUS":
        return RenderableEvent(
            kind=RenderableKind.NOTE,
            title="Agent status",
            severity=Severity.INFO,
            metadata=dict(payload),
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "TASK_STATUS_CHANGED":
        from_status = str(payload.get("from") or "?")
        to_status = str(payload.get("to") or "?")
        return RenderableEvent(
            kind=RenderableKind.STATUS_CHANGE,
            title=f"{from_status} -> {to_status}",
            metadata={"from": from_status, "to": to_status},
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "CRITERION_VERDICT":
        verdict = str(payload.get("verdict") or "")
        is_pass = verdict == "PASS"
        reason = str(payload.get("reason") or "")
        return RenderableEvent(
            kind=RenderableKind.VERDICT,
            title="PASS" if is_pass else "FAIL",
            body=reason,
            severity=Severity.SUCCESS if is_pass else Severity.WARNING,
            metadata={"verdict": verdict, "reason": reason},
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "AGENT_COMPLETED":
        return RenderableEvent(
            kind=RenderableKind.NOTE,
            title="Agent completed",
            severity=Severity.SUCCESS,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "AGENT_FAILED":
        error = str(payload.get("error") or payload.get("details") or "Agent failed")
        return RenderableEvent(
            kind=RenderableKind.ERROR,
            title="Agent failed",
            body=error,
            severity=Severity.ERROR,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "MERGE_COMPLETED":
        return RenderableEvent(
            kind=RenderableKind.MERGE,
            title="Merge completed",
            severity=Severity.SUCCESS,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "MERGE_FAILED":
        error = str(payload.get("error") or "unknown")
        return RenderableEvent(
            kind=RenderableKind.MERGE,
            title="Merge failed",
            body=error,
            severity=Severity.ERROR,
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "PLAN_UPDATE":
        return RenderableEvent(
            kind=RenderableKind.PLAN,
            title="Plan updated",
            event_id=event_id,
            session_id=session_id,
        )

    if event_type == "AUTO_REVIEW_STARTED":
        return RenderableEvent(
            kind=RenderableKind.NOTE,
            title="Auto-review started",
            event_id=event_id,
            session_id=session_id,
        )

    # Unknown event type — return a generic note so callers can decide
    return RenderableEvent(
        kind=RenderableKind.NOTE,
        title=event_type,
        body=str(payload) if payload else "",
        event_id=event_id,
        session_id=session_id,
    )


__all__ = [
    "RenderableEvent",
    "RenderableKind",
    "Severity",
    "render_event",
]

File: packages/vscode/src/api/event-rendering.ts
/**
 * Semantic event rendering — shared protocol for all Kagan clients.
 *
 * Transforms raw `(event_type, payload)` pairs into a client-agnostic
 * `RenderableEvent` that carries kind, title, body, and severity.
 * Mirrors the Python implementation in `kagan.core._event_rendering`.
 */

// ── Enums ────────────────────────────────────────────────────────────────────

export type RenderableKind =
  | "text"
  | "thought"
  | "tool_start"
  | "tool_update"
  | "status_change"
  | "verdict"
  | "note"
  | "error"
  | "merge"
  | "plan";

export type Severity = "info" | "success" | "warning" | "error";

// ── Renderable event ─────────────────────────────────────────────────────────

export interface RenderableEvent {
  kind: RenderableKind;
  title: string;
  body: string;
  severity: Severity;
  metadata: Record<string, unknown>;
  event_id: string;
  session_id: string;
}

// ── Private helpers ──────────────────────────────────────────────────────────

function acpPayload(payload: Record<string, unknown>): Record<string, unknown> {
  const nested = payload.acp;
  return typeof nested === "object" && nested !== null
    ? (nested as Record<string, unknown>)
    : {};
}

/**
 * Format a tool name for human-readable display.
 *
 * - `mcp__kagan__task_get` -> `kagan / task_get`
 * - `toolu_abc` / `call_abc` -> `tool call`
 * - `functions__name` -> `functions / name`
 * - `snake_case` -> `snake case`
 */
export function formatToolName(raw: string): string {
  if (raw.startsWith("toolu_") || raw.startsWith("call_")) return "tool call";
  if (raw.includes("__")) {
    const parts = raw.split("__");
    if ((parts[0] === "mcp" || parts[0] === "functions") && parts.length >= 3) {
      return parts.slice(1).join(" / ");
    }
    return parts.join(" / ");
  }
  return raw.replaceAll("_", " ");
}

/**
 * Extract and format the human-readable tool title from a payload.
 */
export function extractToolTitle(payload: Record<string, unknown>): string {
  const acp = acpPayload(payload);
  const raw = String(
    acp.toolName ??
      acp.name ??
      acp.title ??
      payload.tool_name ??
      payload.toolName ??
      payload.name ??
      payload.tool_call_id ??
      payload.toolCallId ??
      payload.id ??
      "tool call",
  );
  return formatToolName(raw);
}

/**
 * Extract tool execution status from a payload.
 */
export function extractToolStatus(
  payload: Record<string, unknown>,
  fallback: string = "done",
): string {
  const acp = acpPayload(payload);
  return String(acp.status ?? payload.status ?? fallback);
}

// ── Renderable factory helper ────────────────────────────────────────────────

function make(
  kind: RenderableKind,
  title: string,
  opts?: {
    body?: string;
    severity?: Severity;
    metadata?: Record<string, unknown>;
    event_id?: string;
    session_id?: string;
  },
): RenderableEvent {
  return {
    kind,
    title,
    body: opts?.body ?? "",
    severity: opts?.severity ?? "info",
    metadata: opts?.metadata ?? {},
    event_id: opts?.event_id ?? "",
    session_id: opts?.session_id ?? "",
  };
}

// ── Main render function ─────────────────────────────────────────────────────

/**
 * Map a raw event into a {@link RenderableEvent}.
 *
 * Returns `null` for events that should be silently skipped (e.g. a
 * `TOOL_CALL_UPDATE` whose status is `"completed"` or `"done"`).
 */
export function renderEvent(
  eventType: string,
  payload: Record<string, unknown>,
  eventId: string = "",
  sessionId: string = "",
): RenderableEvent | null {
  const ids = { event_id: eventId, session_id: sessionId };

  if (eventType === "OUTPUT_CHUNK") {
    const text = String(payload.text ?? "");
    if (!text) return null;
    const thought = Boolean(payload.thought);
    return make(thought ? "thought" : "text", thought ? "Thinking" : "Output", {
      body: text,
      ...ids,
    });
  }

  if (eventType === "TOOL_CALL_START") {
    return make("tool_start", extractToolTitle(payload), ids);
  }

  if (eventType === "TOOL_CALL_UPDATE") {
    const status = extractToolStatus(payload, "done");
    if (status === "completed" || status === "done") return null;
    return make("tool_update", extractToolTitle(payload), {
      body: status,
      ...ids,
    });
  }

  if (eventType === "AGENT_STATUS") {
    return make("note", "Agent status", {
      metadata: { ...payload },
      ...ids,
    });
  }

  if (eventType === "TASK_STATUS_CHANGED") {
    const from = String(payload.from ?? "?");
    const to = String(payload.to ?? "?");
    return make("status_change", `${from} -> ${to}`, {
      metadata: { from, to },
      ...ids,
    });
  }

  if (eventType === "CRITERION_VERDICT") {
    const verdict = String(payload.verdict ?? "");
    const reason = String(payload.reason ?? "");
    const verdictLabel = verdict === "PASS" ? "PASS" : verdict === "SKIP" ? "SKIP" : "FAIL";
    const verdictSeverity: Severity =
      verdict === "PASS" ? "success" : verdict === "SKIP" ? "info" : "warning";
    return make("verdict", verdictLabel, {
      body: reason,
      severity: verdictSeverity,
      metadata: { verdict, reason },
      ...ids,
    });
  }

  if (eventType === "AGENT_COMPLETED") {
    return make("note", "Agent completed", {
      severity: "success",
      ...ids,
    });
  }

  if (eventType === "AGENT_FAILED") {
    const error = String(payload.error ?? payload.details ?? "Agent failed");
    return make("error", "Agent failed", {
      body: error,
      severity: "error",
      ...ids,
    });
  }

  if (eventType === "MERGE_COMPLETED") {
    return make("merge", "Merge completed", {
      severity: "success",
      ...ids,
    });
  }

  if (eventType === "MERGE_FAILED") {
    const error = String(payload.error ?? "unknown");
    return make("merge", "Merge failed", {
      body: error,
      severity: "error",
      ...ids,
    });
  }

  if (eventType === "PLAN_UPDATE") {
    return make("plan", "Plan updated", ids);
  }

  if (eventType === "AUTO_REVIEW_STARTED") {
    return make("note", "Auto-review started", ids);
  }

  // Unknown event type — return a generic note
  return make("note", eventType, {
    body: Object.keys(payload).length > 0 ? JSON.stringify(payload) : "",
    ...ids,
  });
}
